### PR TITLE
Faster Process.children(recursive=True)

### PR DESCRIPTION
**Pull request by pitrou**
Before:
```
$ python -m timeit -s "import psutil; p = psutil.Process()" "list(p.children(recursive=True))"
10 loops, best of 3: 29.6 msec per loop
```

After:
```
$ python -m timeit -s "import psutil; p = psutil.Process()" "list(p.children(recursive=True))"
100 loops, best of 3: 12.4 msec per loop
```

For reference, non-recursive:
```
$ python -m timeit -s "import psutil; p = psutil.Process()" "list(p.children())"
100 loops, best of 3: 12.2 msec per loop
```

Issue #1183.

### CODE REVIEW

**Comment by pitrou** [comment]
@giampaolo

**Comment by pitrou** [comment]
Basically the idea is that instead of calling `Process` for all processes on the system and then selecting those that are candidate (grand)children of `self`, we prune the search space and only call `Process` for candidate (grand)children.

**Comment by pitrou** [comment]
(there seems to be an unrelated failure on AppVeyor)

File: psutil/__init__.py [hunk 1]
```diff
@@ -888,33 +888,33 @@ def children(self, recursive=False):
...
-                for child in table[pid]:
+                reverse_ppid_map[ppid].append(pid)
+            # Recursively traverse that dict, starting from self.pid,
+            # such that we only call Process() on actual children
+            seen = set()
+            stack = [self.pid]
+            while stack:
+                pid = stack.pop()
+                if pid in seen:
+                    # Since pids can be reused while the ppid_map is
```
Comment (pitrou) [inline psutil/__init__.py]: Note I'm not sure this check is necessary, as the `create_time` test may be sufficient to avoid such potential cycles.

**Comment by giampaolo** [comment]
Thanks, this is awesome. I'll try to take a look at this over the week end.

**Comment by giampaolo** [comment]
Looks good. Merging.

**Pull request merged**